### PR TITLE
Spanner gapic migration error augmentation with interceptor

### PR DESCRIPTION
**Pull request by yihanzhen**
- Add `SpannerInterceptorProvider`
- Move `LoggingInterceptor` outside `GrpcSpannerRpc` and make it a separate class
- Use `SpannerInterceptorProvider` to do logging and error augmentation in `GapicSpannerRpc`

### CODE REVIEW

**Comment by hzyi-google** [comment]
@pongad gax isn't released yet so this PR does not compile now. Can you take a look and see if the overall idea works? I'll release gax afterwards.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java [hunk 1]
```diff
@@ -143,69 +142,69 @@ public GapicSpannerRpc(SpannerOptions options) throws IOException {
...
-                .applyToAllUnaryMethods(
-                    new ApiFunction<UnaryCallSettings.Builder<?, ?>, Void>() {
-                      @Override
-                      public Void apply(UnaryCallSettings.Builder<?,?> builder) {
-                        builder.setRetryableCodes(ImmutableSet.<StatusCode.Code>of());
-                        return null;
-                      }
-                    })
-                .build());
+          GrpcSpannerStub.create(
```
Comment (pongad) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java]: Why this whitespace change, did it not use 2-space before?

Comment (hzyi-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java]: It was wrong before :( (Note that `this.instanceStub` used to have the same indentation as `try {`

Comment (pongad) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java]: In that case, LGTM

**Comment by hzyi-google** [comment]
PTAL @garrettjonesgoogle @snehashah16

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerInterceptorProvider.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * For internal use only.
+ */
+class SpannerInterceptorProvider implements GrpcInterceptorProvider {
+
+  private static final List<ClientInterceptor> clientInterceptors =
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerInterceptorProvider.java]: Would this be in parity with this ?
https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java#L132

- Missing the WatchdogInterceptor

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerInterceptorProvider.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.google.api.gax.grpc.GrpcInterceptorProvider;
+import com.google.common.collect.ImmutableList;
+import io.grpc.ClientInterceptor;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * For internal use only.
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerInterceptorProvider.java]: provide more documentation

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerInterceptorProvider.java (hunk 2 above)
Comment (hzyi-google) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerInterceptorProvider.java]: gax-java itself has watchdog, I'll add it in the next pull request

Comment (hzyi-google) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerInterceptorProvider.java]: Oh sorry didn't see you had two comments. Yes these two are in parity.

**Pull request merged**